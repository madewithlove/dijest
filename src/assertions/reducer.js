import {fromJS} from 'immutable';

/**
 * Assert that a reducer equals X
 */
export default (actual, before, action = {}, after) => {
    before = fromJS(before);
    const result = actual(before, action);

    // If we specified a result to test against
    // use that instead of a snapshot
    if (after) {
        return expect(result.toJS()).toEqual(after);
    }

    return expect(result).toMatchSnapshot();
};
