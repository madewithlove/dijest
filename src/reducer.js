export default function reducer(description, tested, state, action, expected) {
    if (typeof action === 'function') {
        it(description, () => {
            action((...args) => {
                expect(tested(state, ...args)).toMatchSnapshot();
            });
        });
    }

    it(description, () => {
        const result = tested(state, action);

        if (typeof expected !== 'undefined') {
            expect(result).toEqual(expected);
        } else {
            expect(result).toMatchSnapshot();
        }
    });
}
