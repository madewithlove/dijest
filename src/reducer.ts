import { AnyAction, Reducer } from 'redux';

export type Dispatcher = AnyAction | ReducerCallback | null;

export type ReducerCallback = (
    wrapper: (action: AnyAction, ...other: any[]) => any,
) => void;

export type ReducerWrapper = (
    description: string,
    dispatcher: Dispatcher,
    expected?: any,
) => void;

export function reducerTester(tested: Reducer<any>, state: any): ReducerWrapper {
    return (newDescription, dispatcher, expected) =>
        reducer(newDescription, tested, state, dispatcher, expected);
}

export default function reducer(
    description: string,
    tested: Reducer<any>,
    state: any,
    dispatcher?: Dispatcher,
    expected?: any,
): void {
    // If we passed a callback, provide the wrapper to it
    if (typeof dispatcher === 'function') {
        it(description, () => {
            dispatcher((action: AnyAction, ...args) => {
                expect(tested(state, action, ...args)).toMatchSnapshot();
            });
        });
    }

    it(description, () => {
        const result = tested(state, dispatcher as AnyAction);

        if (typeof expected !== 'undefined') {
            expect(result).toEqual(expected);
        } else {
            expect(result).toMatchSnapshot();
        }
    });
}
