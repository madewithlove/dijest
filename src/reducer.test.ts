import reducer from './reducer';

const dummyReducer = (state = 0, { type }) => {
    return type === 'INCREMENT' ? ++state : --state;
};

reducer('can snapshot action', dummyReducer, 0, { type: 'INCREMENT' });
reducer(
    'can snapshot with explicit after state',
    dummyReducer,
    0,
    { type: 'INCREMENT' },
    1,
);
reducer('can do multiple operations', dummyReducer, 0, wrapper => {
    wrapper({ type: 'INCREMENT' });
    wrapper({ type: 'DECREMENT' }, -1);
});
