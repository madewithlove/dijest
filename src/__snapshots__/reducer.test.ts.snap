// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can do multiple of these for a given reducer 1`] = `1`;

exports[`can do multiple of these for a given reducer 2`] = `1`;

exports[`can do multiple of these for a given reducer 3`] = `-1`;

exports[`can snapshot the result of an action 1`] = `1`;

exports[`or even as such 1`] = `1`;

exports[`or even as such 2`] = `1`;

exports[`or even as such 3`] = `-1`;

exports[`which can then be used as such 1`] = `1`;
